README.rst
======================
cookiecutter-pylibrary
======================

Cookiecutter_ template for a Python library.

*Notes*:

* This is largely designed to address this `blog post about packaging python
  libraries <https://blog.ionelmc.ro/2014/05/25/python-packaging/>`_.

  * ... and it will save you from `packaging pitfalls
    <https://blog.ionelmc.ro/2014/06/25/python-packaging-pitfalls/>`_.

* Although this cookiecutter is based off of that of
  `ionelmc <https://github.com/ionelmc/cookiecutter-pylibrary>`_, it also takes
  inspiration from `python-blueprint <https://github.com/johnthagen/python-blueprint/tree/main>`_
  and `cookiecutter-hypermodern-python <https://github.com/cjolowicz/cookiecutter-hypermodern-python/tree/main>`_
* There's a bare library using this template (if you're curious about the final
  result): https://github.com/ugognw/python-no-name.
* If you have a web application (not a library) you might want to take a look at
  `django-docker <https://github.com/evozon/django-docker>`_.

.. contents:: Table of Contents

Features
--------

This is an "all inclusive" sort of template.

* Choice of various licenses.

* Configuration to match your choice of hosting on GitHub or GitLab

* ``src/package_name`` directory structure

* tests outside of package

* Nox_ and nox-poetry_ for managing test environments for PyPy-3.10, Python 3.10, and 3.11

* pytest_ for testing (with `Coverage.py`_ for coverage analysis)

* Automated dependency updates with Dependabot_ (if hosting with GitHub)

* Documentation with Sphinx_, ready for ReadTheDocs_.

* Black_ - for code formating

* Ruff_ for static checks and import sorting

* mypy_ for type-checks to supplement Ruff_

* Coveralls_, Codecov_ Codacy_, and/or `Code Climate`_ integration for coverage tracking

* Virtual environment management and package building/publishing with Poetry_

* CI/CD configuration for testing and building with GitHub Actions or GitLab CI/CD

  * testing code functionality

  * testing documentation builds, links, docstrings

  * checking license compatability (using pip-licenses_) whenever the `pyproject.toml` is changed

  * publishing tagged versions to PyPI (you must store a PyPI token as a secret in order for this to work)

* Version managing with bump2version_

* *Optional* git hooks via pre-commit_

* *Optional* support for testing across different platforms

* *Optional* command-line interface via argparse_, click_, `Python Fire`_, or Typer_ (and rich_)

* Configurations for:

  * bumpversion_ (bump2version_ required)

  * pytest_

  * mypy_

  * pre-commit_

  * Black_

  * `Coverage.py`_

  * Ruff_

  * Dependabot_ or dependabot-standalone_ (not that on GitLab CI/CD, you will have to schedule the Dependabot pipeline; see `here <https://docs.gitlab.com/ee/ci/pipelines/schedules.html>`_)

Requirements
------------

Projects using this template have the following minimal dependencies:

* click_, `Python Fire`_, or Typer_ (and rich_)

To get quickly started on a new system, just `install pip
<https://pip.pypa.io/en/latest/installing.html>`_. That's the bare minimum to required
install Cookiecutter_. To install , just run this in your shell or command prompt::

  pip install cookiecutter

Usage and options
-----------------

This template is more involved than the regular `cookiecutter-pypackage
<https://github.com/audreyr/cookiecutter-pypackage>`_.

First generate your project::

  cookiecutter gh:ugognw/cookiecutter-pylibrary

You will be asked for these fields:

.. note:: Fields that work together usually use the same prefix. If you answer "no" on the first one then the rest
   won't have any effect so just ignore them. Maybe in the future cookiecutter will allow option hiding or something
   like a wizard.

.. list-table::
    :header-rows: 1

    * - Field
      - Default
      - Description

    * - ``full_name``
      - .. code:: python

            "Ugochukwu Nwosu"
      - Main author of this library or application (used in ``AUTHORS.rst`` and ``pyproject.toml``).

        Can be set in your ``~/.cookiecutterrc`` config file.

    * - ``email``
      - .. code:: python

            "ugognw@gmail.com"
      - Contact email of the author (used in ``AUTHORS.rst`` and ``pyproject.toml``).

        Can be set in your ``~/.cookiecutterrc`` config file.

    * - ``website``
      - .. code:: python

            "https://www.law-two.com"
      - Website of the author (used in ``AUTHORS.rst``).

        Can be set in your ``~/.cookiecutterrc`` config file.

    * - ``project_name``
      - .. code:: python

            "Nameless"
      - Verbose project name, used in headings (docs, readme, etc).

    * - ``project_short_description``
      - .. code:: python

            "An example package [...]"
      - One line description of the project (used in ``README.rst`` and ``pyproject.toml``).

    * - ``package_name``
      - .. code:: python

            "nameless"
      - Python package name (whatever you would import via a Python `import` statement).

    * - ``distribution_name``
      - .. code:: python

            "nameless"
      - PyPI distribution name (what you would ``pip install``).

    * - ``repo_name``
      - .. code:: python

            "python-nameless"
      - Repository name on GitHub or GitLab (and project's root directory name).

    * - ``repo_hosting``
      - .. code:: python

            "github.com"
      - You can also use ``"gitlab.com"``. If you desire CI/CD configuration, this
        should be consistent with the values for `github_actions` and `gitlab_ci_cd`.

    * - ``repo_username``
      - .. code:: python

            "ugognw"
      - GitHub or GitLab user name of this project (used for GitHub/GitLab link).

        Can be set in your ``~/.cookiecutterrc`` config file.
    
    * - ``repo_main_branch``
      - .. code:: python
            "main"
      - The name of the default branch for this project.

    * - ``release_date``
      - .. code:: python

            "today"
      - Release date of the project (ISO 8601 format) default to today (used in ``CHANGELOG.rst``).

    * - ``year_from``
      - .. code:: python

            "now"
      - Copyright start year (used in Sphinx ``conf.py``).

    * - ``year_to``
      - .. code:: python

            "now"
      - Copyright end year (used in Sphinx ``conf.py``).

    * - ``keywords``
      - .. code:: python

            "now"
      - List of comma-separated keywords to use in `pyproject.toml` (e.g., `physics,math,chemistry`).

    * - ``version``
      - .. code:: python

            "0.0.1"
      - Release version (see ``.bumpversion.cfg`` and in Sphinx ``conf.py``).

    * - ``license``
      - .. code:: python

            "BSD license"
      - License to use. Available options:

        * BSD license
        * MIT license
        * ISC license
        * Apache Software License 2.0

        What license to pick? https://choosealicense.com/

    * - ``command_line_interface``
      - .. code:: python

            "plain"
      - Option to enable a CLI (a bin/executable file). Available options:

        * ``plain`` - a very simple command.
        * ``argparse`` - a command implemented with argparse_.
        * ``fire`` - a command implemented with `Python Fire`_.
        * ``typer`` - a command implemented with Typer_ (and rich_).
        * ``click`` - a command implemented with click_ - which you can use to build more complex commands.
        * ``no`` - no CLI at all.

    * - ``cli_bin_name``
      - .. code:: python

            "nameless"
      - Name of the CLI bin/executable file (verify that the console script name in
        ``pyproject.toml`` matches your desired implementation; see
        `here <https://python-poetry.org/docs/pyproject/#scripts>`_).

    * - ``pypi_badge``
      - .. code:: python

            "yes"
      - By default, this will insert links to your project's page on PyPI.org.
        Note that if your package is not (yet) on PyPI, this will cause tox -e docs to fail.
        If you choose "no", then these links will not be created.

    * - ``pypi_disable_upload``
      - .. code:: python

            "no"
      - If you specifically want to be sure your package will never be
        accidentally uploaded to PyPI, you can pick "yes".

    * - ``coveralls``
      - .. code:: python

            "yes"
      - Enable pushing coverage data to Coveralls_ and add badge in ``README.rst``. Don't forget
        to add your repo on `https://coveralls.io <https://coveralls.io>`_!

    * - ``codecov``
      - .. code:: python

            "yes"
      - Enable pushing coverage data to Codecov_ and add badge in ``README.rst``. Don't forget
        to add your repo on `https://about.codecov.io <https://about.codecov.io>`_! Note that if
        you opt for CI/CD with GitLab, you must add your Codecov_ API token as a variable called
        ``CODECOV_TOKEN`` (see `here <https://docs.codecov.com/docs/gitlab-2-getting-a-codecov-account-and-uploading-coverage#create-a-codecov-account>`_).

    * - ``codacy``
      - .. code:: python

            "yes"
      - Enable Codacy_ in your chosen CI/CD pipeline and add a corresponding badge in ``README.rst``.
        Don't forget to import your project on `https://www.codacy.com <https://www.codacy.com>`_! 

        **Note:** Displaying the Codacy badge is contingent on your project ID. If you don't
        input your project ID during the cookiecutter configuration step, you can still fill
        in your hexadecimal project ID in the badge URL in the `README.rst`. If you opt for GitLab as your CI/CD platfor, Codacy integration can be enabled by following the steps `here <https://docs.codacy.com/repositories-configure/integrations/gitlab-integration/>`_.

    * - ``codacy_projectid``
      - .. code:: python

            ""
      - Your Codacy_ hexadecimal project ID. Get ID from
        https://app.codacy.com/gh/{{ cookiecutter.repo_username }}/{{ cookiecutter.repo_name }}/settings.

    * - ``codeclimate``
      - .. code:: python

            "yes"
      - Place a `Code Climate`_ badge in the ``README.rst``.

    * - ``gitchangelog``
      - .. code:: python

            "yes"
      - Whether or not to include gitchangelog_ as a dependency.

    * - ``github_actions``
      - .. code:: python

            "yes"
      - Whether or not to use GitHub Actions as your CI/CD framework.

    * - ``gitlab_ci_cd``
      - .. code:: python

            "yes"
      - Whether or not to use GitLab CI/CD as your CI/CD framework.

    * - ``test_on_osx``
      - .. code:: python

            "yes"
      - Whether or not to test your package on OSX in addition to Linux in CI/CD.

    * - ``test_on_windows``
      - .. code:: python

            "yes"
      - Whether or not to test your package on Windows in addition to Linux in CI/CD.

    * - ``pre_commit``
      - .. code:: python

            "yes"
      - Whether or not to enable pre-commit_.

    * - ``install_precommit_hooks``
      - .. code:: python

            "yes"
      - Whether or not to install pre-commit_ hooks. Requires that a .git repository exists in
        the current working directory. If pre-commit_ is not already installed, then it will be
        installed via ``pip``.

    * - ``pytest_datadir``
      - .. code:: python

            "yes"
      - Whether or not to install pytest-datadir_ as a testing dependency.

    * - ``pytest_xdist``
      - .. code:: python

            "yes"
      - Whether or not to install pytest-xdist_ as a testing dependency.

    * - ``sphinx_docs``
      - .. code:: python

            "yes"
      - Have Sphinx documentation.

    * - ``sphinx_theme``
      - .. code:: python

            "furo"
      - What Sphinx_ theme to use.

        Suggested alternative:
        `sphinx-py3doc-enhanced-theme <https://pypi.org/project/sphinx_py3doc_enhanced_theme>`__
        for a responsive theme based on the Python 3 documentation.

    * - ``sphinx_doctest``
      - .. code:: python

            "no"
      - Set to ``"yes"`` if you want to enable doctesting in the `docs` environment. Works best with
        ``test_matrix_separate_coverage == 'no'``.

        Read more about `doctest support in Sphinx <http://www.sphinx-doc.org/en/stable/ext/doctest.html>`_.

    * - ``sphinx_docs_hosting``
      - .. code:: python

            "repo_name.readthedocs.io"
      - Leave as default if your documentation will be hosted on readthedocs.
        If your documentation will be hosted elsewhere (such as GitHub Pages or GitLab Pages),
        enter the top-level URL.

    * - ``initialize_git_repository``
      - .. code:: python

            "yes"
      - Whether or not to initialize a Git repository using `git init`. This also creates an
        initial commit and an initial tag with the version number specified in ``cookiecutter.version``.
        Both are pushed to the repository specified in ``cookiecutter.repo_hosting``.

    * - ``install_package``
      - .. code:: python

            "yes"
      - Whether or not to include install the newly created package via Poetry_.
        If a virtual environment is not already active, this will create a new virtual environment
        in which to install the current package.

Developing the project
``````````````````````

To run all the tests, just run::

  nox

To see all the nox environments::

  nox -l

To only build the docs::

  nox -s docs

To build and verify that the built package is proper and other code QA checks::

  nox -s format,lint

Releasing the project
`````````````````````
Before releasing your package on PyPI you should have all the nox environments passing.

Version management
''''''''''''''''''

This template provides a basic bumpversion_ configuration. It's as simple as running:

* ``bumpversion patch`` to increase version from `1.0.0` to `1.0.1`.
* ``bumpversion minor`` to increase version from `1.0.0` to `1.1.0`.
* ``bumpversion major`` to increase version from `1.0.0` to `2.0.0`.

You should read `Semantic Versioning 2.0.0 <http://semver.org/>`_ before bumping versions.

Building and uploading
''''''''''''''''''''''

Before building dists make sure you got a clean build area::

    rm -rf build
    rm -rf src/*.egg-info

Note:

    Dirty ``build`` or ``egg-info`` dirs can cause problems: missing or stale files in
    the resulting dist or strange and confusing errors. Avoid having them around.

Then you should check that you got no packaging issues::

    nox -s format

And then you can build the ``sdist``, and if possible, the ``bdist_wheel`` too::

    poetry build

To make a release of the project on PyPI, assuming you got some distributions in
``dist/``, the most simple usage is::

    poetry publish

You should set your PyPI credentials according to
`here <https://python-poetry.org/docs/repositories/#configuring-credentials>`_.

Changelog
---------

See `CHANGELOG.rst <https://github.com/ionelmc/cookiecutter-pylibrary/blob/master/CHANGELOG.rst>`_.

FAQs
-------------------

Why is the version stored in several files
(``pkg/__init__.py``, ``pyproject.toml``, ``docs/conf.py``)?

  We cannot use a metadata/version file [#]_ because this template is to be used with
  both distributions of packages (dirs with ``__init__.py``) and modules (simple ``.py``
  files that go straight in ``site-packages``). There's no good place for that extra file
  if you're distributing modules.

  But this isn't so bad - bumpversion_ manages the version string quite
  neatly.

.. [#] Example, an ``__about__.py`` file.

Not Exactly What You Want?
--------------------------

No way, this is the best. :stuck_out_tongue_winking_eye:


If you have criticism or suggestions please open up an Issue or Pull Request.

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _Nox: https://nox.thea.codes/en/stable/
.. _nox-poetry: https://nox-poetry.readthedocs.io/
.. _pytest: http://pytest.org/
.. _Dependabot: https://github.com/dependabot/dependabot-core
.. _dependabot-standalone: https://gitlab.com/dependabot-gitlab/dependabot-standalone
.. _Sphinx: http://sphinx-doc.org/
.. _ReadTheDocs: https://readthedocs.org/
.. _Black: https://black.readthedocs.io/
.. _Ruff: https://beta.ruff.rs/docs/
.. _mypy: https://mypy.readthedocs.io/
.. _pre-commit: https://pre-commit.com
.. _Coverage.py: https://coverage.readthedocs.io/
.. _Coveralls: https://coveralls.io/
.. _Codecov: http://codecov.io/
.. _Codacy: https://codacy.com/
.. _Code Climate: https://codeclimate.com/
.. _Poetry: https://python-poetry.org
.. _pip-licenses: https://github.com/raimon49/pip-licenses
.. _bumpversion: https://pypi.org/project/bump2version
.. _bump2version: https://github.com/c4urself/bump2version
.. _argparse: https://docs.python.org/3/library/argparse.html
.. _click: http://click.pocoo.org/
.. _`Python Fire`: https://github.com/google/python-fire
.. _Typer: https://typer.tiangolo.com
.. _rich: https://rich.readthedocs.io/
.. _gitchangelog: https://github.com/vaab/gitchangelog


CHANGELOG.rst
Changelog
#########

All notable changes to this project will be documented in this file.

The format is based on `Keep a Changelog <https://keepachangelog.com/en/1.0.0/>`_,
and this project adheres to `Semantic Versioning <https://semver.org/spec/v2.0.0.html>`_.

`Unreleased`_
-------------

Added
~~~~~

* Optional Coveralls_, Codecov_, Codacy_, or `Code Climate`_ integration for GitLab or GitHub
* Default GitHub actions and GitLab pipelines for documentation and changelog generation, package publishing, dependency tracking, and license compatability tracking
* Automated dependency updates with Dependabot_ (if hosting with GitHub) or dependabot-standalone_ (if hosting with GitLab)
* Additional CLI engine options:
  * `Python Fire`_ and Typer_ (with rich_)
* Badges in ``README.rst`` for Poetry_, mypy_, `Code Climate`_, nox_
* Support for optional `pytest`_ plugins (pytest_datadir_ and pytest_xdist_)
* Support for optional gitchangelog_
* Support for furo_ documentation theme with Sphinx_ and ReadTheDocs_
* Options to initialize git repository and install the package

Changed
~~~~~~~

* setuptools_ -> Poetry_ for packagement management and building
* Transition from ``setup.py`` to ``pyproject.toml``
* Tox_ -> Nox_ + nox-poetry_ for managing test environments
* Replace Pylint_ with mypy_ for type-checking


Removed
~~~~~~~

* C-extension support
* Optional tests inside of packagement
* Optional coverage testing separate from tests
* Support for hosting domains other than ``github.com`` and ``gitlab.com``
* Removed support for version control via tbump_
* Scrutinizer_ support
* Option for non-Sphinx docs engine

`0.0.1`_ (2023-06-24)
---------------------

Added
~~~~~

* Forked from `ionelmc's version <https://github.com/ionelmc/cookiecutter-pylibrary>`_.

.. _Unreleased: https://github.com/ugognw/cookiecutter-pylibrary/tree/main
.. _`0.0.1`: https://github.com/ugognw/cookiecutter-pylibrary/tree/main
.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _Tox: https://tox.wiki/
.. _Nox: https://nox.thea.codes/en/stable/
.. _nox-poetry: https://nox-poetry.readthedocs.io/
.. _pytest: http://pytest.org/
.. _Dependabot: https://github.com/dependabot/dependabot-core
.. _dependabot-standalone: https://gitlab.com/dependabot-gitlab/dependabot-standalone
.. _Sphinx: http://sphinx-doc.org/
.. _ReadTheDocs: https://readthedocs.org/
.. _Black: https://black.readthedocs.io/
.. _Ruff: https://beta.ruff.rs/docs/
.. _Pylint: http://pylint.readthedocs.io
.. _mypy: https://mypy.readthedocs.io/
.. _Coveralls: https://coveralls.io/
.. _Codecov: http://codecov.io/
.. _Codacy: https://codacy.com/
.. _Code Climate: https://codeclimate.com/
.. _setuptools: http://setuptools.pypa.io
.. _Poetry: https://python-poetry.org
.. _pip-licenses: https://github.com/raimon49/pip-licenses
.. _`Python Fire`: https://github.com/google/python-fire
.. _Typer: https://typer.tiangolo.com
.. _gitchangelog: https://github.com/vaab/gitchangelog
.. _tbump: https://github.com/your-tools/tbump
.. _Scrutinizer: https://scrutinizer-ci.com
.. _rich: https://rich.readthedocs.io/

tests/conftest.py
import datetime
import pytest

from jinja2 import Environment, FileSystemLoader


@pytest.fixture(name='repo_hosting')
def fixture_repo_hosting(request) -> str:
    if request.node.get_closest_marker('repo_hosting'):
        return request.node.get_closest_marker('repo_hosting').args[0]
    else:
        return 'gitlab.com'


@pytest.fixture(name='cookie_config')
def fixture_cookie_config(repo_hosting) -> dict:
    return {
        "full_name": "Ugochukwu Nwosu",
        "email": "ugognw@gmail.com",
        "website": "https://www.law-two.com",
        "project_name": "No Name",
        "project_short_description": "An example package. Generated with cookiecutter-pylibrary.",
        "package_name": "no_name",
        "distribution_name": "no-name",
        "repo_name": "python-no-name",
        "repo_hosting": repo_hosting,
        "repo_username": "ugognw",
        "repo_main_branch": "main",
        "release_date": "today",
        "year_from": str(datetime.datetime.now().year),
        "year_to": str(datetime.datetime.now().year),
        "keywords": "dog,gag",
        "version": "0.0.0",
        "license": "MIT license",
        "command_line_interface": "click",
        "cli_bin_name": "no-name",
        "pypi_badge": "yes",
        "pypi_disable_upload": "no",
        "coveralls": "yes",
        "codecov": "yes",
        "codacy" : "yes",
        "codacy_projectid" : "0000000",
        "codeclimate" : "yes",
        "gitchangelog": "yes",
        "github_actions": "yes",
        "gitlab_ci_cd": "yes",
        "test_on_osx": "yes",
        "test_on_windows": "yes",
        "pre_commit": "yes",
        "install_precommit_hooks": "yes",
        "pytest_datadir": "yes",
        "pytest_xdist": "yes",
        "sphinx_docs": "yes",
        "sphinx_theme": ["furo", "sphinx-rtd-theme", "python-docs-theme", "sphinx-py3doc-enhanced-theme", "sphinx-book-theme", "pydata-sphinx-theme"],
        "sphinx_doctest": "yes",
        "sphinx_docs_hosting": "https://no-name.readthedocs.io/",
        "initialize_git_repository": "yes",
        "install_package": "yes",
        "__repo_url": f"https://{repo_hosting}/ugognw/python-no-name",
    }


@pytest.fixture(name='template_dir')
def fixture_template_dir(request) -> str:
    marker = request.node.get_closest_marker('template_dir')
    if marker:
        return marker.args[0]
    else:
        return '.'


@pytest.fixture(name='environment')
def fixture_environment(template_dir) -> Environment:
    return Environment(
        loader=FileSystemLoader(
            template_dir
        ),
    )
    

tests/test_render_ci_cd.py
from jinja2 import Environment
import pathlib

import pytest


@pytest.fixture(name='template_names')
def fixture_template_names(request):
    if request.node.get_closest_marker('repo_hosting').args[0] == 'github.com':
        return ('licenses.yml', 'publish.yml', 'python.yml')
    elif request.node.get_closest_marker('repo_hosting').args[0] == 'gitlab.com':
        return ('licenses.yml', 'publish.yml', 'python.yml')


class TestGithub:
    @staticmethod
    @pytest.mark.repo_hosting('github.com')
    @pytest.mark.template_dir('{{cookiecutter.repo_name}}/.github/workflows/')
    def test_render(cookie_config: dict, environment: Environment, template_names: list[str], template_dir: str):
        for template_name in template_names:
            filename = pathlib.Path(f'test_results/{template_dir}') / template_name
            template = environment.get_template(template_name)
            with open(filename, mode='w', encoding='utf-8') as file:
                file.write(
                    template.render(
                        cookiecutter=cookie_config
                    )
                )


class TestGitlab:
    @staticmethod
    @pytest.mark.repo_hosting('gitlab.com')
    @pytest.mark.template_dir('{{cookiecutter.repo_name}}/.gitlab/templates/')
    def test_render(cookie_config: dict, environment: Environment, template_names: list[str], template_dir: str):

        for template_name in template_names:
            filename = pathlib.Path(f'test_results/{template_dir}') / template_name
            template = environment.get_template(template_name)
            with open(filename, mode='w', encoding='utf-8') as file:
                file.write(
                    template.render(
                        cookiecutter=cookie_config
                    )
                )


tests/test_render_python_yml.py
from jinja2 import Environment, FileSystemLoader
import pathlib


def test_render_python_yml():
    env = Environment(
        loader=FileSystemLoader(
            '{{cookiecutter.repo_name}}/.github/workflows/'
        ),
    )

    template = env.get_template('python.yml')
    tox_environments = ['py310', 'py311']
    cookiecutter = {
        'sphinx_docs': 'yes',
        'coveralls': 'yes',
        'codecov': 'yes',
        'codacy': 'yes',
        'github_actions_windows': 'yes',
        'github_actions_osx': 'yes'
    }

    filename = pathlib.Path('tests/results') / 'python.yml'

    with open(filename, mode='w', encoding='utf-8') as file:
        file.write(
            template.render(
                cookiecutter=cookiecutter,
                tox_environments=tox_environments
            )
        )


test_render_python_yml()


hooks/post_gen_project.py
import pathlib
import shutil
import subprocess
import sys


try:
    from click.termui import secho
except ImportError:
    warn = note = success = print
else:
    def warn(text):
        for line in text.splitlines():
            secho(line, fg="white", bg="red", bold=True)

    def note(text):
        for line in text.splitlines():
            secho(line, fg="yellow", bold=True)

    def success(text):
        for line in text.splitlines():
            secho(line, fg="green", bold=True)


if __name__ == "__main__":
    cwd = pathlib.Path().resolve()
    src = cwd / 'src'

{%- if 'readthedocs' not in cookiecutter.sphinx_docs_hosting %}
    cwd.joinpath('.readthedocs.yml').unlink()
{% endif %}

{%- if cookiecutter.command_line_interface == 'no' %}
    src.joinpath('{{ cookiecutter.package_name }}', '__main__.py').unlink()
    src.joinpath('{{ cookiecutter.package_name }}', 'cli.py').unlink()
{% endif %}

{%- if cookiecutter.github_actions == 'no' or cookiecutter.repo_hosting != 'github.com' %}
    shutil.rmtree(cwd.joinpath('.github'), ignore_errors=True)
{%- elif cookiecutter.gitlab_ci_cd == 'no' or cookiecutter.repo_hosting != 'gitlab.com' %}
    shutil.rmtree(cwd.joinpath('.gitlab'), ignore_errors=True)
    cwd.joinpath('.gitlab-ci.yml').unlink(missing_ok=True)
{% endif %}

{%- if cookiecutter.codecov == 'yes' %}
    cwd.joinpath('codecov.yml').unlink(missing_ok=True)
{%- endif %}

{%- if cookiecutter.pypi_disable_upload == 'yes' %}
    cwd.joinpath('.github', 'workflows', 'publish.yml').unlink(missing_ok=True)
    cwd.joinpath('.gitlab', 'templates', 'publish.yml').unlink(missing_ok=True)
{%- endif %}

{%- if cookiecutter.license == "no" %}
    cwd.joinpath('LICENSE').unlink()
{%- endif %}

    width = min(140, shutil.get_terminal_size(fallback=(140, 0)).columns)
{%- if cookiecutter.initialize_git_repository == 'yes' %}
    note(' Initializing Git repository '.center(width, "#"))
    try:
        _ = subprocess.check_call(['git', 'init'])
        _ = subprocess.check_call(['git', 'add', '--all'])
        _ = subprocess.check_call(['git', 'commit', '-m', '"Add initial project skeleton."'])
        _ = subprocess.check_call(['git', 'tag', 'v{{ cookiecutter.version }}'])
        _ = subprocess.check_call(['git', 'remote', 'add', 'origin', 'git@{{ cookiecutter.repo_hosting }}:{{ cookiecutter.repo_username }}/{{ cookiecutter.repo_name }}.git'])
        _ = subprocess.check_call(['git', 'push', '-u', 'origin', '{{ cookiecutter.repo_main_branch }}'])
        _ = subprocess.check_call(['git', 'push', '-u', 'origin', 'v{{ cookiecutter.version }}'])
    except subprocess.CalledProcessError as err:
        print(err.args)
{%- endif %}
    package_installed = False
{% if cookiecutter.install_package == 'yes' %}
    note(' Installing package and dependencies '.center(width, "#"))
    try:
        _ = subprocess.check_call(['poetry', 'install'])
        package_installed = True
    except FileNotFoundError as err:
        print(err.args)
        note('Installing poetry'.center(width, "#"))
        p = subprocess.run(
            [
                'curl',
                '-sSL',
                'https://install.python-poetry.org',
                '|',
                'python3',
                '-'
            ], stdout=subprocess.PIPE, shell=True
        )
        _ = subprocess.run(stdin=p.stdout, shell=True)
        note('Poetry successfully installed'.center(width, "#"))
        _ = subprocess.check_call(['poetry', 'install'])
        package_installed = True
    except subprocess.CalledProcessError as err:
        print(err.args)
    except Exception:
        print(err.args)
        warn(
            'Unable to install Poetry.'.center(width, "#")
        )
        warn(
            'You will need to install Poetry in order to install the project and activate the virtual environment.'.center(width, "#")
        )
        warn(
            'https://python-poetry.org/docs/#installation.'.center(width, "#")
        )
{%- endif %}

    pre_commit_installed = False
{%- if cookiecutter.pre_commit == 'no' %}
    cwd.joinpath('.pre-commit-config.yaml').unlink()
{%- elif cookiecutter.install_precommit_hooks == 'yes' %}
    note(' Setting up pre-commit '.center(width, "#"))
    if cwd.joinpath('.git').exists():
        try:
            _ = subprocess.check_call(['poetry', 'run', 'pre-commit', 'install', '--install-hooks'])
            _ = subprocess.check_call(['poetry', 'run', 'pre-commit', 'autoupdate'])
            pre_commit_installed = True
        except FileNotFoundError:
            try:
                _ = subprocess.check_call(['pip', 'install', 'pre-commit'])
            except Exception:
                warn(
                'Unable to install pre-commit.'.center(width, "#")
                )

    else:
        print('Skipping precommit install.')
{%- endif %}

    success(' Successfully created `{{ cookiecutter.repo_name }}` '.center(width, "#"))
    print('See .cookiecutterrc for instructions on regenerating the project.')
    note('To get started run these:')
    commands = [
        'cd {{ cookiecutter.repo_name }}',
{%- if cookiecutter.initialize_git_repository == 'no' %}
        'git init',
{%- endif %}
    ]

{%- if cookiecutter.initialize_git_repository == 'no' %}
    commands.extend(
        (
            'git add --all',
            'git commit -m "Add initial project skeleton."',
            'git tag v{{ cookiecutter.version }}',
            'git remote add origin git@{{ cookiecutter.repo_hosting }}:{{ cookiecutter.repo_username }}/{{ cookiecutter.repo_name }}.git',
            'git push -u origin {{ cookiecutter.repo_main_branch }}',
            'git push -u origin {{ cookiecutter.repo_main_branch }} v{{ cookiecutter.version }}'
        )
    )
{%- endif %}
    if not package_installed:
        commands.append('poetry install')

    commands.append('poetry shell')
{% if cookiecutter.pre_commit == 'yes' %}
    if not pre_commit_installed:
        commands.extend(('pre-commit install --install-hooks', 'pre-commit autoupdate'))
{% endif %}

    print('\n'.join(commands))
    cli_bin_name = '{{ cookiecutter.cli_bin_name }}'
    while cli_bin_name.endswith('.py'):
        cli_bin_name = cli_bin_name[:-3]

        if cli_bin_name == '{{ cookiecutter.package_name }}':
            warn('''
┌───────────────────────────────────────────────────────────────────────┐
│ ERROR:                                                                │
│                                                                       │
│     Your result package is broken. Your bin script named              │
│     {0} │
│                                                                       │
│     Python automatically adds the location of scripts to              │
│     `sys.path`. Because of that, the bin script will fail             │
│     to import your package because it has the same name               │
│     (it will try to import itself as a module).                       │
│                                                                       │
│     To avoid this problem you have two options:                       │
│                                                                       │
│     * Remove the ".py" suffix from `cli_bin_name`. │
│                                                                       │
│     * Use a different `package_name` {1} │
└───────────────────────────────────────────────────────────────────────┘
'''.format(
                '"{{ cookiecutter.cli_bin_name }}" will shadow your package.'.ljust(65),
                '(not "{0}").'.format(cli_bin_name).ljust(32)))
            sys.exit(1)
        break


{{cookiecutter.repo_name}}/noxfile.py
from pathlib import Path
from tempfile import NamedTemporaryFile

import nox
from nox_poetry import Session, session

nox.options.error_on_external_run = True
nox.options.reuse_existing_virtualenvs = True
nox.options.sessions = ['format_check', 'lint', 'type_check', 'test', 'docs']


@session(python=['pypy-3.10', '3.10', '3.11'])
def test(s: Session) -> None:
    s.install('.', 'pytest', 'pytest-cov')
    s.run(
        'python',
        '-m',
        'pytest',
        '--cov=no_name',
        '--cov-report=html',
        '--cov-report=lcov',
        '--cov-report=xml',
        '--cov-report=term-missing',
        'tests',
        *s.posargs,
    )


# For some sessions, set venv_backend="none" to simply execute scripts within the existing Poetry
# environment. This requires that nox is run within `poetry shell` or using `poetry run nox ...`.
@session(venv_backend='none')
def format(s: Session) -> None:
    s.run('ruff', 'check', '.', '--select', 'I', '--fix')
    s.run('black', '.')


@session(venv_backend='none')
def format_check(s: Session) -> None:
    s.run('ruff', 'check', '.', '--select', 'I')
    s.run('black', '--check', '.')


@session(venv_backend='none')
def lint(s: Session) -> None:
    s.run('ruff', 'check', '.')


@session(venv_backend='none')
def lint_fix(s: Session) -> None:
    s.run('ruff', 'check', '.', '--fix')


@session(venv_backend='none')
def type_check(s: Session) -> None:
    s.run('mypy', 'src', 'tests', 'noxfile.py')


@session(venv_backend='none')
def docs(s: Session) -> None:
    s.run(
        'sphinx-build',
        '--color',
        '-W',
        '-b',
        'html',
        'docs/source',
        'docs/build',
    )


@session(venv_backend='none')
def docs_check_urls(s: Session) -> None:
    s.run('sphinx-build', '--color', '-W', '-b', 'linkcheck')


{%- if cookiecutter.sphinx_doctest == 'yes' %}
@session(venv_backend='none')
def docs_test(s: Session) -> None:
    s.run('sphinx-build', '--color', '-W', '-b', 'doctest')
{%- endif %}


# Note: This reuse_venv does not yet have affect due to:
#   https://github.com/wntrblm/nox/issues/488
@session(reuse_venv=False)
def licenses(s: Session) -> None:
    # Generate a unique temporary file name. Poetry cannot write to the temp file directly on
    # Windows, so only use the name and allow Poetry to re-create it.
    with NamedTemporaryFile() as t:
        requirements_file = Path(t.name)

    # Install dependencies without installing the package itself:
    #   https://github.com/cjolowicz/nox-poetry/issues/680
    s.run_always(
        'poetry',
        'export',
        '--without-hashes',
        f'--output={requirements_file}',
        external=True,
    )
    s.install('pip-licenses', '-r', str(requirements_file))
    s.run("pip-licenses",
          '--from=all',
          '--no-version',
          '--format=html',
          '--output-file=licenses-summary.html',
          *s.posargs)
    requirements_file.unlink()
    